Source/Core/VideoBackends/Software/SWVertexLoader.cpp

```diff
@@ -134,76 +131,90 @@ static T ReadNormalized(I value)
   return casted;
 }
 
-template <typename T, bool swap = false>
-static void ReadVertexAttribute(T* dst, DataReader src, const AttributeFormat& format,
-                                int base_component, int components, bool reverse)
+template <typename T>
+static void ReadVertexAttribute0(T* dst, DataReader src, const AttributeFormat& format,
+                                 std::size_t base_component, std::size_t components, bool reverse)
 {
-  if (format.enable)
-  {
-    src.Skip(format.offset);
-    src.Skip(base_component * (1 << (format.type >> 1)));
+  ASSERT(format.components >= base_component);
 
-    int i;
-    for (i = 0; i < std::min(format.components - base_component, components); i++)
-    {
-      int i_dst = reverse ? components - i - 1 : i;
-      switch (format.type)
-      {
-      case VAR_UNSIGNED_BYTE:
-        dst[i_dst] = ReadNormalized<T, u8>(src.Read<u8, swap>());
-        break;
-      case VAR_BYTE:
-        dst[i_dst] = ReadNormalized<T, s8>(src.Read<s8, swap>());
-        break;
-      case VAR_UNSIGNED_SHORT:
-        dst[i_dst] = ReadNormalized<T, u16>(src.Read<u16, swap>());
-        break;
-      case VAR_SHORT:
-        dst[i_dst] = ReadNormalized<T, s16>(src.Read<s16, swap>());
-        break;
-      case VAR_FLOAT:
-        dst[i_dst] = ReadNormalized<T, float>(src.Read<float, swap>());
-        break;
-      }
-
-      ASSERT_MSG(VIDEO, !format.integer || format.type != VAR_FLOAT,
-                 "only non-float values are allowed to be streamed as integer");
-    }
-    for (; i < components; i++)
+  src.Skip(format.offset);
+  const std::size_t component_size = (1ull << (format.type >> 1));
+  src.Skip(base_component * component_size);
+
+  for (std::size_t i = 0; i < std::min(format.components - base_component, components); i++)
+  {
+    const std::size_t i_dst = reverse ? components - i - 1 : i;
+    switch (format.type)
     {
-      int i_dst = reverse ? components - i - 1 : i;
-      dst[i_dst] = i == 3;
```

Comment:
Note: this loop (which applies to the components not affected by the previous loop) is somewhat confusing, but I think the intention is to supply default values for uninitialized components (e.g. if only XY are provided, it gives a default for Z, and if only RGB are provided, it gives a default for A).  As far as I can tell, it sets the 4th component (`i == 3`, which becomes `dst[0]` when reverse is enabled) to 1 and the other uninitialized ones to 0, i.e. it's trying to set the alpha component to 1 for colors.  However, colors use a `u8`, so if it were trying to set it to the max value, it should use 255 instead.

I've changed the relevant code to use `std::fill`/`std::fill_n` and just always sets 0, which doesn't seem to have made a difference; I haven't hardware tested what actually happens in that case though.

I don't think that case could even be hit, because `format.components` is always set to 4 for colors ([here](https://github.com/dolphin-emu/dolphin/blob/c77a5f7e3246398aeb1d520d7bbc0bfdf5787754/Source/Core/VideoCommon/VertexLoader.cpp#L146)/[here](https://github.com/dolphin-emu/dolphin/blob/c77a5f7e3246398aeb1d520d7bbc0bfdf5787754/Source/Core/VideoCommon/VertexLoaderX64.cpp#L467)).  But this code is very weird looking so it's worth mentioning it.